### ISSUE TITLE
Site Permission settings on Focus

### ISSUE THREAD

**Issue by amedyne** [mozilla-mobile/focus-android#6504]
Provide users with settings and options for Site Permissions.
See design on Figma [here](https://www.figma.com/file/6fLq7hHGcaFlhunnlhAVCk/Site-permissions-Update?node-id=2522%3A7249).

Users should have settings for

- [x] Autoplay
- [x] Camera
- [x] Location
- [x] Microphone
- [x] Notification
- [x] DRM-controlled content

**Comment by amedyne** [comment]
Tagging @jeffreygee to provide setting options for each Setting and the default values.

**Comment by jeffreygee** [comment]
@amedyne - Let's go with the same settings options for each item as what we have on Fenix

Suggested defaults (feel free to confirm @channingcarter) 
- Autoplay - Block Audio only
- Camera - Ask to allow 
- Location - Blocked by Android 
- Microphone - Blocked by Android 
- Notification - Ask to allow
- DRM-controlled content - Ask to allow

**Comment by Mugurell** [comment]
@jeffreygee @channingcarter In regards to how these permissions would apply there are 3 scenarios:
- app-wide - them and any of their change applying to all websites. Would the users always have to go to settings to make a change / check the permissions status or will the permissions will also be shown in the toolbar like on Fenix?
Having to go in settings to a specific section might be too much for all users and there might be some who would not know about that and report issues about website features not working.
- app wide but allow for temporary exceptions - when on a website users will have the possibility (from the toolbar like on Fenix?) to change a certain permission for just that browsing instance, with no data about this persisted. At every new website access they would have to modify the permission again if they want something different than the default.
-  app wide but allow for permanent exceptions (this is the behavior on Fenix) - when on a website users will have the possibility (from the toolbar like on Fenix?) to change a certain permission for that specific website with the app remembering this change and as such remember user's browsing history.

This is the current UX on Fenix for adding permanent site permissions exceptions (which can then be edited in app settings):



https://user-images.githubusercontent.com/11428869/155675303-11026f60-40df-4228-b986-9110aa565776.mp4

**Comment by channingcarter** [comment]
@Mugurell I'm fine with mimicking Fenix in regards to be able to manage site permissions from the toolbar. However, we should not remember a change from a certain site because as you indicated that is remembering a user's browsing history which goes against the intent of Focus.

**Comment by Mugurell** [comment]
Talked with Gabriel a bit on how this should all work:
Following the behavior on Fenix also our proposal is:

This ticket will add the defaults from https://github.com/mozilla-mobile/focus-android/issues/6504#issuecomment-1050424136 to app settings.
  - These settings can only be changed by the user from app settings
  - These settings will apply to all tabs and webpages
  
This implies:
- If the default in app settings is "Ask to allow" everytime a website tries to use a permission the user will be asked whether to allow it or not. Their response will not be persisted in any way. The default from app settings will remain "Ask to allow".
- If the default is "Allowed" then everytime a website tries to use a permission it will be allowed to without any user interaction or disruption of the browsing flow.  (Other than first asking for the system hardware permission if not already granted)
We could show the blinking dot indicator in the toolbar to get user's focus on a permission's view there. Can be part of another ticket.
- If the default is "Blocked" then everytime a website tries to use a permission it will be blocked without any user interaction or disruption of the browsing flow. 
We could show the blinking dot indicator in the toolbar to get user's focus on a permission's view there. Can be part of another ticket.
- While the system hardware permission is not granted for a complementary app permission we will show a message similar to the one from Fenix:
<img src="https://user-images.githubusercontent.com/11428869/157883965-3e012476-6fe7-424f-8ab6-4f3134fbee5a.png" width="33%" />

- https://github.com/mozilla-mobile/focus-android/issues/6504#issuecomment-1050612472 can be a followup allowing users to change a particular permission for just that website then and there. The default from app settings won't be changed. Nothing will be persisted.


Does this sound okay? @jonalmeida @jeffreygee  ?

**Comment by Amejia481** [comment]
> We could show the blinking dot indicator in the toolbar to get user's focus on a permission's view there. Can be part of another ticket.

This is the same that we have on Fenix, I believe we could get the behavior for from [AC](https://github.com/mozilla-mobile/android-components/blob/743081d2dc85630dae9e0e23904f067dfcde47bf/components/feature/sitepermissions/src/main/java/mozilla/components/feature/sitepermissions/SitePermissionsFeature.kt#L365) for free, we only need to activate 
[Indicators.HIGHLIGHT](https://github.com/mozilla-mobile/android-components/blob/743081d2dc85630dae9e0e23904f067dfcde47bf/components/browser/toolbar/src/main/java/mozilla/components/browser/toolbar/display/DisplayToolbar.kt#L82) :)

**Comment by Amejia481** [comment]
> Talked with Gabriel a bit on how this should all work: Following the behavior on Fenix also our proposal is:This ticket will add the defaults from [#6504 (comment)](https://github.com/mozilla-mobile/focus-android/issues/6504#issuecomment-1050424136) to app settings.
> 
>     * These settings can only be changed by the user from app settings
> 
>     * These settings will apply to all tabs and webpages
> 
> 
> This implies:
> 
>     * If the default in app settings is "Ask to allow" everytime a website tries to use a permission the user will be asked whether to allow it or not.  Their response will not be persisted in any way. The default from app  settings will remain "Ask to allow".
> 
>     * If the default is "Allowed" then everytime a website tries to use a  permission it will be allowed to without any user interaction or  disruption of the browsing flow. (Other than first asking for the system hardware permission if not already granted)
>       We could show the  blinking dot indicator in the toolbar to get user's focus on a  permission's view there. Can be part of another ticket.
> 
>     * If the default is "Blocked" then everytime a website tries to use a  permission it will be blocked without any user interaction or disruption of the browsing flow.
>       We could show the blinking dot indicator in  the toolbar to get user's focus on a permission's view there. Can be  part of another ticket.
> 
>     * While the  system hardware permission is not granted for a complementary app  permission we will show a message similar to the one from Fenix:
> 
> 
> ![img](https://user-images.githubusercontent.com/11428869/157883965-3e012476-6fe7-424f-8ab6-4f3134fbee5a.png)
> 
>     * [Site Permission settings on Focus #6504 (comment)](https://github.com/mozilla-mobile/focus-android/issues/6504#issuecomment-1050612472) can be a followup allowing users to change a particular permission for  just that website then and there. The default from app settings won't be changed. Nothing will be persisted.
> 
> 
> Does this sound okay? @jonalmeida @jeffreygee ?

This sounds good to me 👍🏽

**Comment by Amejia481** [comment]
@Mugurell 

I'm bit worry about the ask to allow option, as there are some permissions prompts, that will include a check-box to store the user choice, I'm don't think we have API that allow to customize that. I will be good to double check.

**Comment by Mugurell** [comment]
> @Mugurell
> 
> I'm bit worry about the ask to allow option, as there are some permissions prompts, that will include a check-box to store the user choice, I'm don't think we have API that allow to customize that. I will be good to double check.

Looks like this can be customized through [`SitePermissionsDialogFragment#newInstance#shouldShowDoNotAskAgainCheckBox`](https://github.com/mozilla-mobile/android-components/blob/ac74a577b68be269a7d86ed753d176d6d1c33a45/components/feature/sitepermissions/src/main/java/mozilla/components/feature/sitepermissions/SitePermissionsDialogFragment.kt#L201-L232) which indeed [defaults to true](https://github.com/mozilla-mobile/android-components/blob/ac74a577b68be269a7d86ed753d176d6d1c33a45/components/feature/sitepermissions/src/main/java/mozilla/components/feature/sitepermissions/SitePermissionsDialogFragment.kt#L71-L72) so this should be set everytime as `false` from Focus.
Definitely need to be careful about it, thank you!

**Comment by Amejia481** [comment]
Yeah, thanks for taking a look. We will need a way to expose it through `SitePermissionsFeature`  :)

**Comment by iorgamgabriel** [comment]
@jeffreygee I think for Camera the default option should be "Blocked by Android" because we need the user permission from android system settings. The same value like Location and Microphone.  In Fenix is also "Blocked by Android" for Camera.

**Comment by jeffreygee** [comment]
@iorgamgabriel - agreed, let's go with that

**Comment by lobontiumira** [comment]
Verified on the latest Focus Nightly 101.0a1 from 4/7 that the Site Permissions are implemented and functioning.
Here are my findings:
- the Camera, Location, Microphone, and Notification options have the same sub-page, and set by default as blocked by the OS, 
- tapping on "Go to Settings" button from the Camera, Location, Microphone, and Notification options, redirects the user to the   settings of the device - Apps - Focus,
- the websites don't "remember" if the user allows or denies the permissions, so the prompt is displayed every time a permission is requested,
- changing the permissions toggles from the device's settings, deletes the opened tabs.

Below are some screenshots of the UI of the Site permissions options, and prompt:

![site permission](https://user-images.githubusercontent.com/40694158/162208140-259355e9-a8c2-4585-8e27-88a73ab6c0a5.png)
